### PR TITLE
Update `interfaces` module to use `bind_new_parameters`

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the effort towards `QuantumScript` immutability. In the end we want to deprecate `QuantumScript.set_parameters` altogether.

**Description of the Change:**
Define a new function `bind_new_parameters_tape` that takes a tape and the new parameters, and returns a new tape. The original tape is unmodified.

Update the `jax`, `jax-jit`, `tensorflow`, and `tensorflow-autograph` interfaces to call this new `bind_new_parameters_tape` function instead of `QuantumScript.set_parameters`.

### CODE REVIEW

File: pennylane/pulse/parametrized_evolution.py
Comment (mudit2812) [inline pennylane/pulse/parametrized_evolution.py:510]: Why not put this inside `ops/functions/bind_new_parameters.py`? Is it because that would create a circular dependency?

File: pennylane/tape/qscript.py
Comment (mudit2812) [inline pennylane/tape/qscript.py:1428]: Can you add `indices` to the docstring as well?

File: pennylane/tape/qscript.py
Comment (mudit2812) [reply pennylane/tape/qscript.py:1459]: Yep. Since `Operator` calls `_check_batching` in `__init__`, this line is redundant.
```suggestion
```

File: pennylane/tape/qscript.py
Comment (mudit2812) [inline pennylane/tape/qscript.py:1462]: Same here
```suggestion
```

**Review by mudit2812 — COMMENTED** [review]
Some redundant-code related suggestions, but overall looks good! Happy to approve once comments are addressed.

File: pennylane/pulse/parametrized_evolution.py
Comment (eddddddy) [reply pennylane/pulse/parametrized_evolution.py:510]: Yep, because `ParametrizedEvolution` isn't available top-level.

File: pennylane/tape/qscript.py
Comment (timmysilv) [inline pennylane/tape/qscript.py:1453]: by only checking existence in `indices`, we don't care about order. I'm not _totally_ opposed to that, but perhaps we should make it clear. Example:
```pycon
>>> qs = qml.tape.QuantumScript([qml.RX(1.1, 0), qml.RY(2.2, 1)])
>>> qs.bind_new_parameters([3.3, 4.4], [0, 1]).circuit
[RX(3.3, wires=[0]), RY(4.4, wires=[1])]
>>> qs.bind_new_parameters([3.3, 4.4], [1, 0]).circuit
[RX(3.3, wires=[0]), RY(4.4, wires=[1])]
```
We could instead follow today's typical strategy of a `trainable_only` boolean toggle, setting indices to `tape.trainable_params if trainable_only else range(len(tape._par_info))`, since that seems to be the only 2 cases. That said, we'd lose some generality

File: pennylane/interfaces/jax_jit.py
Comment (eddddddy) [reply pennylane/interfaces/jax_jit.py:1]: Yep, should be fine since they would be removed in the same release.

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py:1453]: This behaviour is the same as `set_parameters`, see the test `test_setting_parameters_unordered` in `test_tape.py`.

I also considered a `trainable_only` flag, but as you said it loses some generality. One example is for the broadcasting transforms where the indices correspond to parameters containing a batch dimension.

File: tests/tape/test_qscript.py
Comment (mudit2812) [inline tests/tape/test_qscript.py:252]:
```suggestion
        batch_size when creating a QuantumScript."""
```

**Review by mudit2812 — APPROVED** [review]
Looks good! I don't think that the discussion about `bind_new_parameters_tape` vs `tape.bind_new_parameters` affects approval.

**Comment by mudit2812** [comment]
I also just realized once this PR is merged, we should probably go back to the gradients module to update the changes from #4220 to use `QuantumScript.bind_new_parameters` instead. It virtually does the same thing.

**Comment by eddddddy** [comment]
Yep, and also with the broadcasting transforms. That's the intention.

File: pennylane/tape/qscript.py
Comment (mudit2812) [inline pennylane/tape/qscript.py:1444]: I just realized that it might be possible to speed up this approach. Rather than iterating through the entire circuit, it makes more sense to just iterate through the indices for which the parameters are being updated. Currently we have O(n) time complexity with n being the number of items in the circuit, and the proposed approach would have complexity O(i) with i being the number of parameters to update. See [here](https://github.com/PennyLaneAI/pennylane/blob/330491ece7d8a2cf2db238c8cfd508bfd7378ae7/pennylane/gradients/general_shift_rules.py#L384) in `qml.gradients`.

**Review by mudit2812 — COMMENTED** [review]
See comment.

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py:1444]: Hmmm, that makes sense. I can't use `get_operation` here though since that only gets trainable parameters, but there's definitely a way to make this work by using `tape._par_info` directly.

Comment (eddddddy) [reply pennylane/tape/qscript.py]: Updated 👍

File: pennylane/tape/qscript.py
Comment (mudit2812) [inline pennylane/tape/qscript.py:1444]:
```suggestion
        _, op_idx, p_idx = tape._par_info[idx]
```

**Review by mudit2812 — APPROVED** [review]
Looks great! Thanks Edward :)

File: pennylane/tape/qscript.py
Comment (timmysilv) [reply pennylane/tape/qscript.py:1444]: this won't work, `tape._par_info["idx"]` is a dict, and it has 3 keys

File: pennylane/tape/qscript.py
Comment (albi3ro) [reply pennylane/tape/qscript.py:1478]: So this functionality should be either a function or a method, not both. 

Pros of function:
* Easier to hide away

Pros of method:
* we don't have to import the function, as it already exists on the object.

We should also modify the same in such a way that it is clearly distinguished from `bind_new_parameters`, in that it can accept a specific subset of indices. But `bind_new_parameters_at_indices` is too long. 🤔 

Also, `bind_new_parameters` itself will be able to be redone following the completion of pytrees. `bind_new_parameters_at_indices` will not.

OK, I'm way overthinking this.

1) Let's just have it as a method, not a standalone function.
2) Let's just call it `bind_new_parameters`, even though it differs from `functions.bind_new_parameters`

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py:1444]: Yep, this won't work.

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py]: Updated 👍

File: pennylane/tape/qscript.py
Comment (timmysilv) [reply pennylane/tape/qscript.py:774]: also, do we need to update the new tape for execution? Perhaps we can add the `_update=True` flag (and interfaces internally set it to False) like @albi3ro is doing in other PRs

**Review by timmysilv — COMMENTED** [review]
a few little concerns left, but otherwise lgtm!

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py]: We at least need to update the tape batch size, so yes. Isn't `update=True` already the default option?

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py:778]: I was also wondering about this. If we eventually want tapes to be immutable, then we really shouldn't be modifying the new tape attributes in-place, right? Attributes like `_qfunc_output` aren't very important and it doesn't matter if they're set, but the `trainable_params` attribute definitely is, and we don't currently have a way to set it without in-place modification.

We do this in a lot of other places as well, so it's probably fine at the moment. It's just something else to keep in mind.

Comment (timmysilv) [reply pennylane/tape/qscript.py:778]: still not sure but we can talk about it outside of this PR. but to clarify as an aside, `_qfunc_output` is _very_ important for 2 things:
- reshaping outputs after qnode execution
- reshaping outputs in legacy gradient computation

I don't think the qnode's tape will go through this function, so it's less important. But until the old return system is removed, the second point is still important.

Comment (eddddddy) [reply pennylane/tape/qscript.py:778]: I see 🤔. But yah this is not the place to discuss this.

File: pennylane/tape/qscript.py
Comment (eddddddy) [reply pennylane/tape/qscript.py]: Whether it needs updating depends on the parameters, but no I don't think it's necessary for the interfaces. I can add a keyword argument then.

Comment (eddddddy) [reply pennylane/tape/qscript.py]: OK I was wrong. The tape needs to be updated in order for `tape._par_info` to be populated, otherwise it's just an empty list.

**Review by timmysilv — APPROVED** [review]
the update upon re-construction is disappointing but I see that it must be done. looks good to me!

**Comment by eddddddy** [comment]
[sc-41633]

**Pull request merged**
